pkg/dns/dnsprovider/dnsProvider.go

```diff
@@ -0,0 +1,57 @@
+package dnsprovider
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/apis/v1alpha1"
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/dns"
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/dns/aws"
+)
+
+var errUnsupportedProvider = fmt.Errorf("provider type given is not supported")
+
+type providerFactory struct {
+	client.Client
+}
+
+func NewProvider(c client.Client) *providerFactory {
+
+	return &providerFactory{
+		Client: c,
+	}
+}
+
+// depending on the provider type specified in the form of a custom secret type https://kubernetes.io/docs/concepts/configuration/secret/#secret-types in the dnsprovider secret it returns a dnsprovider.
+func (p *providerFactory) DNSProviderFactory(ctx context.Context, managedZone *v1alpha1.ManagedZone) (dns.Provider, error) {
+	providerSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      managedZone.Spec.SecretRef.Name,
+			Namespace: managedZone.Spec.SecretRef.Namespace,
+		}}
+
+	if err := p.Client.Get(ctx, client.ObjectKeyFromObject(providerSecret), providerSecret); err != nil {
+		return nil, err
+	}
+
+	switch providerSecret.Type {
+	case "kuadrant.io/aws":
+		log.Log.Info("Creating DNS provider for provider type AWS")
+		DNSProvider, err := aws.NewProviderFromSecret(providerSecret)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create dns provider from secret: %v", err)
+		}
+		log.Log.Info("Route53 provider created")
+
+		return DNSProvider, nil
```

Comment:
Since this gets called all over the place, you might want to consider what logs you want here, currently you get: 
```
2023-07-12T12:15:36+01:00       INFO    Creating DNS provider for provider type AWS
2023-07-12T12:15:36+01:00       INFO    Route53 provider created
```

repeated a lot.

Maybe a single log that mentions the managed zone it's being create for might be more useful , and possibly a debug log instead of info.

Reply:
Ok cool ill change that